### ISSUE TITLE
Wav files played with Bass change pitch and duration 

### ISSUE THREAD

**Issue by jsm174** [vpinball/vpinball#224]
When using Bass to play wav files (Standalone builds), the pitch and duration seem to change. It is very noticeable when rapidly pressing the flippers.

Here is an example:

https://user-images.githubusercontent.com/1197137/208246368-4d382223-15ad-48bb-9748-27e99ae34672.mp4

I know Bass is not used for wav files in the regular builds, but I just wanted to document this behavior.

**Comment by Scottacus64** [comment]
In addition to this when bass encounters a WAV file that is longer in duration than the timer that is driving it, bass will not start an additional instance of the sound file.  For example a 162ms timer runs five times to score 50 points in Grand Tour (Bally EM where the score motor scores a set of five 10 point scores) and the sound file for the score motor run is around 174ms long, you will only hear the first, third and fifth sounds played.  If the file is shortened to 150ms, all five score motor sounds play.  The native vpx audio player has no problem with this.   Also the ball rolling sounds stutter when played with bass.

**Comment by jsm174** [comment]
@Scottacus64 and myself spent some time on this.

I noticed a comment 

`// Frequency tweaks are relative to original sound.  If the copy failed for some reason, don't alter original`

When going through the Bass route, it was loading the last frequency:

```
         float freq;
         BASS_ChannelGetAttribute(m_BASSstream, BASS_ATTRIB_FREQ, &freq);
```

Every time you press a flipper, the frequency grows and grows. 

So, now I cache the frequency when the stream is created and then re-use it `Play`:

https://github.com/vpinball/vpinball/blob/04a6df923ac24180ac60641d45316cf07a783c0a/pinsound.cpp#L269-L274

I also attempt to re-work the loop / restart logic to match the DirectX version.

DX:

https://github.com/vpinball/vpinball/blob/04a6df923ac24180ac60641d45316cf07a783c0a/pinsound.cpp#L856-L861

New BASS:

https://github.com/vpinball/vpinball/blob/04a6df923ac24180ac60641d45316cf07a783c0a/pinsound.cpp#L331-L339


I don't think is correct as it still seems a little off.

**Comment by Scottacus64** [comment]
Jsm174 has fixed the ball rolling sound problem as well.  It seems to me that the issue is with how the playsound command and its parameters are being used by bass.  The problem seems to revolve around the "useexisting" and "repeat" parameters.  As I understand it useexisting is set to 1 for ball rolling and repeat is set to 0 and that seems to work well in jsm174's new code.  Sounds like my score motor sets useexisting to 1 and restart to 1 as well.  Those sounds are behaving as if useexisting is set to 1 and repeat is set to 0 (a sound file that is playing will play till the end and not restart again until finished playing to the end).  The score motor sound file is longer than the timer that triggers it so if five score motor sounds are played by the timer, only 1,3,5 are heard while 2,4 are blocked by the playing of 1,3.

**Comment by toxieainc** [comment]
I would love to be of greater help here, but as you may have already noticed, the BASS code path was not completely finished back then, especially the WAV path. I don't know what other stuff distracted me back then ;) , but i fear you guys are now more the experts on this than me.

**Comment by Scottacus64** [comment]
It turns out that jsm174's code is perfect, there was a problem with the vpx table script where restart was set to 0 in the playsound command so BASS was not terminating the first sound so the second could play as in this code snippet.
`DWORD status; 
 m_pDSBuffer->GetStatus(&status); 
 if (!(status & DSBSTATUS_PLAYING)) 
 	m_pDSBuffer->Play(0, 0, flags); 
 else if (restart) 
 	m_pDSBuffer->SetCurrentPosition(0); `

**Comment by toxieainc** [comment]
Is this port above then enough to close this issue or is there more outstanding? Thanks @jsm174 and @Scottacus64 !

**Comment by jsm174** [comment]
Yes the only thing that I'm really not 100% sure about is `useexisting`.  TBH, I'm lost when stepping through it.

The `ONLY_USE_BASS` flow never does anything like `PinDirectSoundWavCopy`.

I'm not sure if this is an issue since it does seem to sound fine now.

**Comment by vbousquet** [comment]
I have been doing some testing with latest (10.8) and it looks to me that the BASS implementation is now fully functional. 

Therefore, I think it would be nice to:
- make it the default sound path in 10.8 (i.e. define ONLY_USE_BASS),
- advertise about it with the 10.8 release so that table authors drop the legacy 'music' folder usage and use the ogg file format (for fairly smaller file size, like what was done regarding Webp/Exr for textures).

This would make this code path more thoroughly tested (by all users) and ease the standalone merge we should have in the future by preparing to drop the legacy path without troubles.

@toxieainc @Scottacus64 @jsm174 Are you aware of any missing functionality/bug that would make this a bad choice ?

**Comment by jsm174** [comment]
I think BASS is working pretty well now. Standalone uses BASS for everything: Music player, PinMAME audio streaming, and faking the WMP.ocx COM player. 

One comment on music folder...   10.8 introduced `ScriptGlobalTable::get_MusicDirectory`, which works great because of all the cross-platform path separator issues. Tables like Blood Machines just happened to use a variable name called `MusicDirectory` which caused some issues.

If there were any changes I would wish for, it would be nice if `MusicDirectory` supported an optional parameter for a sub folder.

So for example `MusicDirectory "Blood Machines OST"` which could then properly return:

`/Users/jmillard/.vpinball/music/Blood Machines OST/`

this way, no path separator code would be needed at all and it could really clean up alot of the VBS scripts.

**Comment by Scottacus64** [comment]
Is there a possibility in BASS to map the sound output directly to the speakers and not use a room simulation?  Cab owners with speakers in the corners under the playfield could get accurate sound placement with this.  It would eliminate any cross talk between the playfield and the backglass sounds that currently happen with the use of the room simulation.

**Comment by toxieainc** [comment]
@nkissebe would hopefully know. He revamped the multi-speaker mapping a while back.

**Comment by vbousquet** [comment]
I just made BASS the only used API for the next beta roll (I should push a beta 3 soon since most bugs are now solved). We will see if there are any report from a broader use of it.

I also added an optional parameter to MusicDirectory (see: https://github.com/vpinball/vpinball/pull/563/commits/33ec6cbdc33050fc73c971d3e0b8f34d9b620923).

**Comment by nkissebe** [comment]
Its been a while and don't work on Windows professionally. So I am hesitant to suggest anything. My Enhanced SSF mode was made to address @scottacus64 observation about the separation of sounds from the backglass and it was 100% when I left it, but that doesn't mean nothing changed. I used BASS a little with SSFImpactor in DOF and you can direct audio per speaker but for SSF you need the sound blended between speakers (sound from the center comes out of all four playfield speakers for instance, or at least we try to). There is probably a way to do it, I was only ever familiar with a tiny bit of the API.. not even enough to really do everything I wanted with SSFImpactor (I had problems figuring out how to make different sound files overlap/mix when two sound effects happen at same time). 

Whatever changes are being made here I hope get well tested for SSF positioning effects for mechanical sounds (which were always mono WAV files, at least when I left it).

**Comment by vbousquet** [comment]
Closing as sound code has been moved to SDL / MiniAudio
